File: static/loyalty.js
let loyaltyData = { loyaltyMembers: [], rewards: [] };
let currentMember = null;
let selectedReward = null;

loadLoyaltyData();

function loadLoyaltyData() {
  fetch("/api/data")
    .then(response => response.json())
    .then(data => {
      loyaltyData = data;
      renderRewards();

      const defaultMember = loyaltyData.loyaltyMembers[0];
      if (defaultMember) {
        document.getElementById("memberIdInput").value = defaultMember.id;
        document.getElementById("lastNameInput").value = defaultMember.lastName;
        currentMember = defaultMember;
        renderMemberProfile(currentMember);
        renderActivityHistory(currentMember);
      }

      bindLoyaltyEvents();
    });
}

function bindLoyaltyEvents() {
  document.getElementById("lookupMemberBtn").onclick = lookupMember;

  document.getElementById("redeemForm").onsubmit = function (e) {
    e.preventDefault();

    if (!currentMember || !selectedReward) {
      alert("Please look up a member and select a reward.");
      return;
    }

    fetch("/api/loyalty/redeem", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        memberId: currentMember.id,
        rewardId: selectedReward.id,
        reservationId: document.getElementById("reservationId").value.trim(),
        notes: document.getElementById("redemptionNotes").value.trim(),
        date: getTodayLabel()
      })
    })
      .then(response => response.json())
      .then(result => {
        if (!result.success) {
          alert(result.message || "Unable to process redemption.");
          return;
        }
        alert("Redemption processed.");
        selectedReward = null;
        highlightSelectedReward();
        reloadMember(currentMember.id, currentMember.lastName);
        document.getElementById("redeemForm").reset();
      });
  };

  document.getElementById("addPointsBtn").onclick = function () {
    if (!currentMember) {
      alert("Please look up a member first.");
      return;
    }

    fetch("/api/loyalty/add-points", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        memberId: currentMember.id,
        points: 500,
        date: getTodayLabel()
      })
    })
      .then(response => response.json())
      .then(result => {
        if (result.success) {
          alert("500 points added.");
          reloadMember(currentMember.id, currentMember.lastName);
        }
      });
  };
}

function lookupMember() {
  const lookup = document.getElementById("memberIdInput").value.trim();
  const lastName = document.getElementById("lastNameInput").value.trim();

  fetch(`/api/loyalty/member?lookup=${encodeURIComponent(lookup)}&lastName=${encodeURIComponent(lastName)}`)
    .then(response => response.json())
    .then(data => {
      if (!data.member) {
        currentMember = null;
        document.getElementById("memberProfileContainer").innerHTML =
          `<div class="message-box">No member found.</div>`;
        document.getElementById("activityHistoryContainer").innerHTML = "";
        return;
      }

      currentMember = data.member;
      renderMemberProfile(currentMember);
      renderActivityHistory(currentMember);
    });
}

function reloadMember(memberId, lastName) {
  fetch(`/api/loyalty/member?lookup=${encodeURIComponent(memberId)}&lastName=${encodeURIComponent(lastName)}`)
    .then(response => response.json())
    .then(data => {
      if (data.member) {
        currentMember = data.member;
        renderMemberProfile(currentMember);
        renderActivityHistory(currentMember);
      }
    });
}

function renderMemberProfile(member) {
  const progressPercent = Math.min((member.availablePoints / member.nextTierPoints) * 100, 100);
  const remainingPoints = Math.max(member.nextTierPoints - member.availablePoints, 0);

  document.getElementById("memberProfileContainer").innerHTML = `
    <div class="member-card">
      <div class="member-tier">✦ ${member.tier}</div>
      <div class="member-name">${member.fullName}</div>
      <div class="member-meta">Member ID: ${member.id} · Enrolled: ${member.enrolled}</div>
      <div class="member-stats">
        <div>
          <div class="member-stat-label">Available Points</div>
          <div class="member-stat-value">${formatNumber(member.availablePoints)}</div>
        </div>
        <div>
          <div class="member-stat-label">Lifetime Stays</div>
          <div class="member-stat-value small-white">${member.lifetimeStays}</div>
        </div>
        <div>
          <div class="member-stat-label">Nights This Year</div>
          <div class="member-stat-value small-white">${member.nightsThisYear}</div>
        </div>
      </div>
    </div>
    <div class="progress-wrap">
      <div class="progress-header">
        <span>Progress to ${member.nextTier}</span>
        <span>${formatNumber(member.availablePoints)} / ${formatNumber(member.nextTierPoints)} pts</span>
      </div>
      <div class="progress-bar">
        <div class="progress-fill" style="width:${progressPercent}%"></div>
      </div>
      <div class="progress-labels">
        <span>${member.tier.replace(" Member", "")}</span>
        <span>${formatNumber(remainingPoints)} pts to ${member.nextTier}</span>
        <span>${member.nextTier}</span>
      </div>
    </div>
  `;
}

function renderActivityHistory(member) {
  if (!member.activityHistory || member.activityHistory.length === 0) {
    document.getElementById("activityHistoryContainer").innerHTML = `<div class="message-box">No activity history available.</div>`;
    return;
  }

  const rows = member.activityHistory.map(item => `
    <tr>
      <td>${item.date}</td>
      <td>${item.description}</td>
      <td>${item.refId}</td>
      <td>${item.points >= 0 ? "+" : ""}${formatNumber(item.points)}</td>
      <td>${formatNumber(item.balance)}</td>
    </tr>
  `).join("");

  document.getElementById("activityHistoryContainer").innerHTML = `
    <table class="activity-table">
      <thead>
        <tr>
          <th>Date</th>
          <th>Description</th>
          <th>Ref ID</th>
          <th>Points</th>
          <th>Balance</th>
        </tr>
      </thead>
      <tbody>${rows}</tbody>
    </table>
  `;
}

function renderRewards() {
  const rewardGrid = document.getElementById("rewardGrid");
  rewardGrid.innerHTML = "";

  loyaltyData.rewards.forEach(reward => {
    const card = document.createElement("div");
    card.className = "reward-card";
    card.dataset.rewardId = reward.id;
    card.innerHTML = `
      <div class="reward-title">${reward.title}</div>
      <div class="reward-points">${formatNumber(reward.points)} pts</div>
      <div class="reward-desc">${reward.description}</div>
    `;
    card.onclick = function () {
      selectedReward = reward;
      highlightSelectedReward();
    };
    rewardGrid.appendChild(card);
  });
}

function highlightSelectedReward() {
  document.querySelectorAll(".reward-card").forEach(card => {
    card.classList.toggle("selected", selectedReward && card.dataset.rewardId === selectedReward.id);
  });
}

function formatNumber(value) {
  return Number(value).toLocaleString();
}

function getTodayLabel() {
  const now = new Date();
  const month = now.toLocaleString("en-US", { month: "short" });
  const day = String(now.getDate()).padStart(2, "0");
  const year = now.getFullYear();
  return `${month} ${day} ${year}`;
}

File: loyalty.py
from database import get_connection


def fetch_loyalty_members():
    conn = get_connection()
    members = conn.execute("SELECT * FROM loyalty_members ORDER BY full_name").fetchall()
    result = []

    for member in members:
        activity = conn.execute("""
            SELECT date, description, ref_id, points, balance, notes
            FROM loyalty_activity
            WHERE member_id = ?
            ORDER BY id DESC
        """, (member["id"],)).fetchall()

        result.append({
            "id": member["id"],
            "email": member["email"],
            "lastName": member["last_name"],
            "firstName": member["first_name"],
            "fullName": member["full_name"],
            "tier": member["tier"],
            "enrolled": member["enrolled"],
            "availablePoints": member["available_points"],
            "lifetimeStays": member["lifetime_stays"],
            "nightsThisYear": member["nights_this_year"],
            "nextTier": member["next_tier"],
            "nextTierPoints": member["next_tier_points"],
            "activityHistory": [
                {
                    "date": a["date"],
                    "description": a["description"],
                    "refId": a["ref_id"],
                    "points": a["points"],
                    "balance": a["balance"],
                    "notes": a["notes"]
                } for a in activity
            ]
        })

    conn.close()
    return result


def fetch_rewards():
    conn = get_connection()
    rows = conn.execute("SELECT * FROM rewards ORDER BY title").fetchall()
    conn.close()
    return [dict(row) for row in rows]


def lookup_member(lookup_value, last_name):
    members = fetch_loyalty_members()
    lookup_lower = lookup_value.strip().lower()
    last_name_lower = last_name.strip().lower()

    member = next(
        (
            m for m in members
            if (m["id"].lower() == lookup_lower or m["email"].lower() == lookup_lower)
            and m["lastName"].lower() == last_name_lower
        ),
        None
    )
    return {"member": member}


def redeem_reward(data):
    member_id = data.get("memberId", "")
    reward_id = data.get("rewardId", "")
    reservation_id = data.get("reservationId", "") or "RDM-MANUAL"
    notes = data.get("notes", "")

    conn = get_connection()
    member = conn.execute("SELECT * FROM loyalty_members WHERE id = ?", (member_id,)).fetchone()
    reward = conn.execute("SELECT * FROM rewards WHERE id = ?", (reward_id,)).fetchone()

    if not member or not reward:
        conn.close()
        return {"success": False, "message": "Member or reward not found."}

    if member["available_points"] < reward["points"]:
        conn.close()
        return {"success": False, "message": "Not enough points."}

    new_balance = member["available_points"] - reward["points"]

    conn.execute("""
        UPDATE loyalty_members
        SET available_points = ?
        WHERE id = ?
    """, (new_balance, member_id))

    conn.execute("""
        INSERT INTO loyalty_activity (member_id, date, description, ref_id, points, balance, notes)
        VALUES (?, ?, ?, ?, ?, ?, ?)
    """, (
        member_id,
        data.get("date", ""),
        f"Reward Redemption - {reward['title']}",
        reservation_id,
        -reward["points"],
        new_balance,
        notes
    ))

    conn.commit()
    conn.close()
    return {"success": True}


def add_points_manually(data):
    member_id = data.get("memberId", "")
    points_to_add = int(data.get("points", 0))

    conn = get_connection()
    member = conn.execute("SELECT * FROM loyalty_members WHERE id = ?", (member_id,)).fetchone()

    if not member:
        conn.close()
        return {"success": False, "message": "Member not found."}

    new_balance = member["available_points"] + points_to_add

    conn.execute("""
        UPDATE loyalty_members
        SET available_points = ?
        WHERE id = ?
    """, (new_balance, member_id))

    conn.execute("""
        INSERT INTO loyalty_activity (member_id, date, description, ref_id, points, balance, notes)
        VALUES (?, ?, ?, ?, ?, ?, ?)
    """, (
        member_id,
        data.get("date", ""),
        "Manual points adjustment",
        "MANUAL-ADD",
        points_to_add,
        new_balance,
        ""
    ))

    conn.commit()
    conn.close()
    return {"success": True}